php-frontend/src/main/java/org/sonar/php/parser/PHPGrammar.java

```diff
@@ -1299,6 +1301,33 @@ public ExpressionTree LIST_ELEMENT() {
         LIST_EXPRESSION()));
   }
 
+  public AssignmentExpressionTree ARRAY_DESTRUCTURING_ASSIGNMENT() {
+    return b.<AssignmentExpressionTree>nonterminal(PHPLexicalGrammar.ARRAY_DESTRUCTURING_ASSIGNMENT).is(
+      f.arrayDestructuringAssignment(ARRAY_ASSIGNMENT_PATTERN(), b.token(EQU), EXPRESSION()));
+  }
+
+  public ArrayAssignmentPatternTree ARRAY_ASSIGNMENT_PATTERN() {
+    return b.<ArrayAssignmentPatternTree>nonterminal(Kind.ARRAY_ASSIGNMENT_PATTERN)
+      .is(f.arrayAssignmentPattern(
+        b.token(LBRACKET),
+        ARRAY_ASSIGNMENT_PATTERN_ELEMENT(),
+        b.zeroOrMore(f.newTuple1(b.token(COMMA), ARRAY_ASSIGNMENT_PATTERN_ELEMENT())),
+        b.token(RBRACKET)));
+  }
+
+  public Tree ARRAY_ASSIGNMENT_PATTERN_ELEMENT() {
+    return b.<Tree>nonterminal(PHPLexicalGrammar.ARRAY_ASSIGNMENT_PATTERN_ELEMENT)
+      .is(
+        b.firstOf(
+          f.arrayAssignmentPatternElement(
+            b.optional(
+              f.newTuple6(
+                EXPRESSION(),
+                b.token(DOUBLEARROW))),
+            VARIABLE_IDENTIFIER()),
```

Comment:
VARIABLE_IDENTIFIER is too restrictive. It should include:
* variables, ex: [$a]
* indexed array, ex: [$a[$i]]
* object property, ex:[$a->name]

This is some valide php code that SonarPHP doesn't parse:
```
$t1 = [3, 5, 7];
$t2 = [9, 8, 6];
$i=2;
[$t1[0], $t1[$i]] = $t2; // indexed array
var_dump($t1);
```
Output:
```
[9, 5, 8]
```

Other example:
```
$t2 = [9, 8, 6];
class A { public $age = 12; }
$a = new A();
[$a->age] = $t2; // object property
var_dump($a);
```
Output:
```
object(A) {  age => 9 }
```